pennylane/tape/tape.py

```diff
@@ -872,6 +879,267 @@ def set_parameters(self, params, trainable_only=True):
+                        if shots != 1:
+                            shape.append(tuple([shots, len(mps)]))
+                        else:
+                            shape.append((len(mps),))
+
+        return shape
+
+    def get_output_shape(self, device):
+        """Produces the output shape of the tape by inspecting its measurements
+        and the device used for execution.
+
+        Note: as the output shape may be dependent on the device used for
+        execution, tapes do not store the computed shape.
+
+        Args:
+            device (~.Device): the device that will be used for the tape execution
+
+        Raises:
+            UnsupportedTapeOperationError: raised for unsupported cases for
+                example when the tape contains heterogeneous measurements
+
+        Returns:
+            Union[tuple[int], list[tuple[int]]]: the output shape(s) of the
+            tape result
+        """
+        output_shape = tuple()
+
+        if len(self._measurements) == 1:
+            output_shape = self._single_measurement_shape(self._measurements[0], device)
+        else:
+            num_measurements = len(set(meas.return_type for meas in self._measurements))
+            if num_measurements == 1:
+                output_shape = self._multi_homogenous_measurement_shape(self._measurements, device)
+            else:
+                raise UnsupportedTapeOperationError(
+                    "Getting the output shape of a tape that contains multiple types of measurements is unsupported."
+                )
+        return output_shape
+
+    def get_output_domain(self):
+        """Returns the numeric type corresponding to the output domain of the
+        tape by inspecting its measurements.
+
+        This function can be used to determine the dtpe of the tape output
+        results before executing the tape.
+
+        Raises:
+            UnsupportedTapeOperationError: raised for unsupported cases for
+                example when the tape contains heterogeneous measurements
+
+        Returns:
+            type: the numeric type corresponding to the output domain of the
+            tape
+        """
+        output_domain = float
+
+        for observable in self._measurements:
+            ret_type = observable.return_type
+            if ret_type == qml.operation.State:
+                return complex
+
+            if ret_type == qml.operation.Sample:
+
+                if observable.obs is None or all(
+                    np.issubdtype(e.dtype, int) for e in observable.eigvals
+                ):
+                    # qml.sample() or integer eigvals
+                    output_domain = int
+                else:
+                    output_domain = float
+
+                # Note: if one of the sample measurements contains outputs that
+                # are real, then the entire result will be real
+                if output_domain == float:
+                    return output_domain
```

Comment:
@antalszava this strikes me as something we should also simply have in the measurement, e.g.,

```python
Measurement(..., shape=(2,), dtype=np.float64)
```

Reply:
Added a `numeric_type` attribute.